### ISSUE TITLE
A problem with loss computation.

### ISSUE THREAD

**Issue by yxdr** [keon/seq2seq#23]
loss = F.nll_loss(output[1:].view(-1, vocab_size), trg[1:].contiguous().view(-1), ignore_index=pad)

The loss computed by the above line is the average at every time step, which can cause it difficult to train the model. 
So I suggest accumulating the loss at every time step. In my experiments, this makes it easier to train the model.

**Comment by fengxin619** [comment]
> loss = F.nll_loss(output[1:].view(-1, vocab_size), trg[1:].contiguous().view(-1), ignore_index=pad)
> 
> The loss computed by the above line is the average at every time step, which can cause it difficult to train the model.
> So I suggest accumulating the loss at every time step. In my experiments, this makes it easier to train the model.

so ,how to write the loss?

**Comment by keon** [comment]
`F.nll_loss(..., reduction='mean')` with `ignore_index=pad` averages NLL over non-pad target tokens in the batch — not 'per timestep'. This is the standard cross-entropy formulation: it's what `nn.CrossEntropyLoss` defaults to and what fairseq, OpenNMT, and the official PyTorch seq2seq tutorial all use.

Sum reduction makes loss magnitude (and gradient scale) depend on sequence-length distribution within a batch, which usually hurts more than it helps once you also have padding. If training is unstable, the productive knobs are LR / dropout / gradient clipping / LR schedule — PR #31 adds `ReduceLROnPlateau` for this reason.

Closing as won't-fix.
